plugin/metrics/setup.go

```diff
@@ -49,7 +50,22 @@ func setup(c *caddy.Controller) error {
 		})
 		return nil
 	})
+	c.OnStartup(func() error {
+		conf := dnsserver.GetConfig(c)
+		plugins := conf.Handlers()
+		for _, h := range conf.ListenHosts {
+			addr, err := net.ResolveTCPAddr("tcp", net.JoinHostPort(h, conf.Port))
+			if err != nil {
+				return err
+			}
+			addrstr := conf.Transport + "://" + addr.String()
+			for _, p := range plugins {
+				vars.PluginEnabled.WithLabelValues(addrstr, conf.Zone, p.Name()).Set(1)
+			}
+		}
+		return nil
 
+	})
 	c.OnRestart(m.OnRestart)
```

Comment:
Did you try with the `reload` plugin ?
And then you change Corefile => CoreDNS will be reloaded .. with maybe different set of plugins ...
Will your metrics show this new set of plugins ?

Reply:
I think you should take advantage of the "onRestart" event. But I am not sure how you can "clean" the metric values of the plugins that were loaded on firt time.

Reply:
> Did you try with the reload plugin ?
And then you change Corefile => CoreDNS will be reloaded .. with maybe different set of plugins ...
Will your metrics show this new set of plugins ?

I tried just now, and it doesn't work:
1. Every time I added a new plugin, after CoreDNS was reloaded, it will show the new set of plugins, but if I removed some plugins, it still showed them.
2. This mertic can not monitor if the `reload` plugin is enabled, which means I can not see `reload` even if I have `metrics` plugin already enabled... I don't really know why.

Reply:
> I think you should take advantage of the "onRestart" event. But I am not sure how you can "clean" the metric values of the plugins that were loaded on firt time.

Good point, I will try to play with that.

Reply:
> This mertic can not monitor if the reload plugin is enabled, which means I can not see reload even if I have metrics plugin already enabled... I don't really know why.

I guess that `reload` does not register as a `handler` because it does not act on the DNS request: there is no `ServerDNS` function in the `reload` plugin.

Reply:
> Every time I added a new plugin, after CoreDNS was reloaded, it will show the new set of plugins, but if I removed some plugins, it still showed them.

I guess you need to find a way to 'clean" the metrics variable on each reload, and re-register the whole new list of plugins.
=> I see a function `Reset()` on a `metricsMap` that seems to be a part of any metrics ...

Reply:
> I guess that reload does not register as a handler because it does not act on the DNS request: there is no ServerDNS function in the reload plugin.

So I just ignore the `reload` plugin?

> I see a function Reset() on a metricsMap that seems to be a part of any metrics ...

Actually, this works fine, I updated the `OnRestart` callback.

Reply:
> So I just ignore the reload plugin?

I think we should fix that on a separated PR for the plugins that does not register a Handler.
So far, I think : `reload` and `bind`.
Best, here, would be to add a section `Bug` at the end of this readme, adding something like:
"Only the plugins that register as Handler are visible in this metrics. As of today the plugins `reload` and `bind` will not be reported" (or something similar).

> Actually, this works fine, I updated the OnRestart callback.

Great !!!
 ... but. I just remind now, that there will be an issue if the Restart fails. Which means if the new Corefile is invalid. In that case, your var is Reset, and CoreDNS will continue running w/o any plugin reported. (because onRestart is called ONLY at starting of CoreDNS).

There is a way to fix that : use the event "onRestartFailed" and let him do exactly the same thing as "onRestart".  See usage of event described in Caddy:
https://github.com/mholt/caddy/blob/f6e6a6be0446a8455001d6202032b20b64b37569/caddy.go#L110-L116


NOTE: [related issue to use onRestartFailed to ensure proper config if the Corefile is invalid at reload](https://github.com/coredns/coredns/issues/2659).

Reply:
> I think we should fix that on a separated PR for the plugins that does not register a Handler.

I am willing to work on this, and I will create an issue for this, any ideas about how to fix this?

> Best, here, would be to add a section Bug at the end of this readme, adding something like:
"Only the plugins that register as Handler are visible in this metrics. As of today the plugins reload and bind will not be reported" (or something similar).

Great, I will do this.

> There is a way to fix that : use the event "onRestartFailed" and let him do exactly the same thing as "onRestart". See usage of event described in Caddy:

Did you mean that the "onRestartFailed" should do the same thing as **`onStart`** since the `onRestart` just resets the metric. We shouldn't reset the metric if it restarts failed, right?

Reply:
Sorr, YES, I meant `onRestartFailed` should do the same thing as `onStart`.

Reply:
Hi @fturib, I tried to do the same thing in `onRestartFailed` as `onStart`, but it doesn't work, I think it is beccause:
 [If that new server fails to start, then the initial server instance is still available and DNS queries still served, **but Prometheus handler stays down. Prometheus will not reply HTTP request until a successful reload or a complete restart of CoreDNS.**](https://github.com/coredns/coredns/blob/master/plugin/metrics/README.md#bugs)

This has been already reported in the `metrics` README.md, so I think we should fix this at first.

Reply:
You are right. The issue opened is here : #2659.